From a9a4cd0c1e09cb3db72c568a4abae74c85a6cc14 Mon Sep 17 00:00:00 2001
From: Zhantao Tang <zhantao.tang@windriver.com>
Date: Sat, 8 Mar 2025 02:25:03 +0000
Subject: [PATCH] uboot: secure: update secboot codes to compatible with SDK
 BSP43

Since the uboot has been updated to SDK BSP43 release, and the secure
boot flow also changed a lot, so the hse_secboot_enable command should
be modified to keep align with latest status.

And the codes in this commit takes the hse secboot implementation in
pkcs11-hse repo as reference.

The main changes are:
 1. add blx key settings to satisfy new secboot flow
 2. new functions to access and set these new blx keys
 3. add environment variables like blx_pubkey, bl2_bin, bl2_sign to
    specify the resources needed for secboot
 4. new functions to access and set these new environment variables
 5. other settings or changes

Upstream-Status: Inappropriate [WR Linux specific]

Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 arch/arm/mach-s32/s32-cc/hse_adv_secboot.c | 506 ++++++++++++++-------
 include/configs/s32-cc.h                   |   5 +
 include/configs/s32g274ardb2.h             |   2 +
 include/configs/s32g2xxaevb.h              |   2 +
 include/configs/s32g399ardb3.h             |   2 +
 include/configs/s32g3xxaevb.h              |   2 +
 include/hse/hse_abi.h                      |  49 +-
 7 files changed, 403 insertions(+), 165 deletions(-)

diff --git a/arch/arm/mach-s32/s32-cc/hse_adv_secboot.c b/arch/arm/mach-s32/s32-cc/hse_adv_secboot.c
index 1e0483ff..09235a94 100644
--- a/arch/arm/mach-s32/s32-cc/hse_adv_secboot.c
+++ b/arch/arm/mach-s32/s32-cc/hse_adv_secboot.c
@@ -25,41 +25,39 @@ DECLARE_GLOBAL_DATA_PTR;
 
 /* hse nvm key catalog configuration */
 #define HSE_NVM_KEY_CATALOG_CFG \
-{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, \
-	HSE_KEY_TYPE_AES, 5U, HSE_KEY128_BITS }, \
-{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, \
-	HSE_KEY_TYPE_AES, 10U, HSE_KEY256_BITS }, \
-{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, \
-	HSE_KEY_TYPE_HMAC, 5U, HSE_KEY512_BITS }, \
-{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, \
-	HSE_KEY_TYPE_ECC_PAIR, 2U, HSE_KEY256_BITS }, \
-{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, \
-	HSE_KEY_TYPE_ECC_PUB, 2U, HSE_KEY256_BITS }, \
-{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, \
-	HSE_KEY_TYPE_ECC_PUB_EXT, 1U, HSE_KEY256_BITS }, \
-{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, \
-	HSE_KEY_TYPE_RSA_PAIR, 2U, HSE_KEY2048_BITS}, \
-{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, \
-	HSE_KEY_TYPE_RSA_PUB, 2U, HSE_KEY2048_BITS}, \
-{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, \
-	HSE_KEY_TYPE_RSA_PUB_EXT, 1U, HSE_KEY2048_BITS}, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_AES, 5U, HSE_KEY128_BITS }, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_AES, 10U, HSE_KEY256_BITS }, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_HMAC, 5U, HSE_KEY512_BITS }, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_ECC_PAIR, 1U, HSE_KEY256_BITS }, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_ECC_PUB, 1U, HSE_KEY256_BITS }, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_RSA_PUB, 1U, HSE_KEY2048_BITS}, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_RSA_PAIR, 2U, HSE_KEY2048_BITS}, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_RSA_PUB, 1U, HSE_KEY2048_BITS}, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_AES, 5U, HSE_KEY256_BITS }, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_HMAC, 5U, HSE_KEY512_BITS }, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_ECC_PAIR, 1U, HSE_KEY256_BITS }, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_ECC_PUB, 1U, HSE_KEY256_BITS }, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_RSA_PAIR, 2U, HSE_KEY2048_BITS}, \
+{ HSE_ALL_MU_MASK, HSE_KEY_OWNER_CUST, HSE_KEY_TYPE_RSA_PUB, 2U, HSE_KEY2048_BITS}, \
 { 0U, 0U, 0U, 0U, 0U }
 
 /* hse ram key catalog configuration */
 #define  HSE_RAM_KEY_CATALOG_CFG \
-{HSE_ALL_MU_MASK, HSE_KEY_OWNER_ANY, \
-	HSE_KEY_TYPE_AES, 6u, HSE_KEY256_BITS }, \
-{HSE_ALL_MU_MASK, HSE_KEY_OWNER_ANY, \
-	HSE_KEY_TYPE_HMAC, 6u, HSE_KEY512_BITS}, \
-{HSE_ALL_MU_MASK, HSE_KEY_OWNER_ANY, \
-	HSE_KEY_TYPE_AES, 7u, HSE_KEY256_BITS }, \
-{HSE_ALL_MU_MASK, HSE_KEY_OWNER_ANY, \
-	HSE_KEY_TYPE_SHARED_SECRET, 1u, HSE_KEY256_BITS}, \
+{HSE_ALL_MU_MASK, HSE_KEY_OWNER_ANY, HSE_KEY_TYPE_AES, 4u, HSE_KEY256_BITS }, \
+{HSE_ALL_MU_MASK, HSE_KEY_OWNER_ANY, HSE_KEY_TYPE_HMAC, 4u, HSE_KEY512_BITS}, \
+{HSE_ALL_MU_MASK, HSE_KEY_OWNER_ANY, HSE_KEY_TYPE_AES, 7u, HSE_KEY256_BITS}, \
+{HSE_ALL_MU_MASK, HSE_KEY_OWNER_ANY, HSE_KEY_TYPE_SHARED_SECRET, 1u, HSE_KEY256_BITS}, \
+{HSE_ALL_MU_MASK, HSE_KEY_OWNER_ANY, HSE_KEY_TYPE_HMAC, 3u, HSE_KEY512_BITS}, \
 {0u, 0u, 0u, 0u, 0u}
 
-#define APP_CODE_OFFSET 0x40
-#define M7_BOOT_CODE_SIZE 0x2000
-#define M7_SIGNATURE_SIZE 0x100
+struct bl_key secboot_keys[] = {
+	{ BL2_KEY, true, BL2_DEFAULT_KEY_HANDLE, NULL, "bl2_pubkey", NULL },
+	{ BL31_KEY, true, BL31_DEFAULT_KEY_HANDLE, NULL, "bl31_pubkey", NULL },
+	{ BL32_KEY, false, BL32_DEFAULT_KEY_HANDLE, NULL, "bl32_pubkey", NULL },
+	{ BL33_KEY, true, BL33_DEFAULT_KEY_HANDLE, NULL, "bl33_pubkey", NULL },
+	{ INVALID_KEY, false, HSE_INVALID_KEY_HANDLE, NULL, NULL, NULL } /* end of key list - required */
+};
+
 
 #define UUID_BL2_CERT \
 	{ 0xea69e2d6, \
@@ -82,11 +80,11 @@ static struct hse_smr_dynamic_config smr_dynamic_config;
 
 static struct spi_flash *qspi_flash;
 
-static int hse_read(u32 offset, size_t len, void *buf, bool qspi_boot)
+static int hse_read(u32 offset, size_t len, void *buf, enum medium_type boot_medium)
 {
 	int ret = 0;
 
-	if (qspi_boot)
+	if (boot_medium == QSPI)
 		ret = spi_flash_read(qspi_flash, offset, len, buf);
 	else
 		ret = hse_mmc_read(buf, offset/SD_SECTOR_SIZE, len/SD_SECTOR_SIZE);
@@ -94,7 +92,7 @@ static int hse_read(u32 offset, size_t len, void *buf, bool qspi_boot)
 	return ret;
 }
 
-static int hse_write(u32 offset, size_t len, const char *buf, bool qspi_boot)
+static int hse_write(u32 offset, size_t len, const char *buf, enum medium_type boot_medium)
 {
 	int ret = 0;
 	uint32_t rem, div;
@@ -102,7 +100,7 @@ static int hse_write(u32 offset, size_t len, const char *buf, bool qspi_boot)
 	const char *end = buf + len;
 	size_t todo;
 
-	if (qspi_boot) {
+	if (boot_medium == QSPI) {
 		if (offset + len  > qspi_flash->size) {
 			printf("ERROR: hse_write failed because of exceeding flash end !\n");
 			return -1;
@@ -142,6 +140,159 @@ static int hse_write(u32 offset, size_t len, const char *buf, bool qspi_boot)
 	return ret;
 }
 
+int get_env_keys()
+{
+
+	char *value = NULL;
+	/* the keys for secboot include: bl2_pubkey, bl31_pubkey, bl32_pubkey, bl33_pubkey and m7_pubkey,
+	 * in which, the bl2, bl31 and bl33 keys are required and the bl32 and m7 keys are optional,
+	 * moreover, for each key, it has a handle, which indicates the key catlog/slot to store the key.
+	 * the key format is:
+	 *     <key_name>:<handle_value>
+	 * by default, all the keys use same value and handle as bl2 key's.
+	 * Users can set their specific values.
+	*/
+
+	for (int i = 0; secboot_keys[i].type != INVALID_KEY; i++) {
+
+		/* get key configs, the config includes pubkey name and the related key handle, the delimiter is ":" */
+		value = env_get(secboot_keys[i].env_name);
+		if (!value || !(*value)) {
+			if (secboot_keys[i].req) {
+				printf("ERROR: the %s is required, but not found\n", secboot_keys[i].env_name);
+				return -ENOENT;
+			}
+			continue;
+		}
+		secboot_keys[i].config = value;
+	}
+
+	return 0;
+}
+
+int hse_parse_keys(void)
+{
+	char *handle_string = NULL;
+
+	for (int i = 0; secboot_keys[i].type != INVALID_KEY; i++) {
+		if (!secboot_keys[i].config)
+			continue;
+
+		if (!strchr(secboot_keys[i].config, ':')) {
+			secboot_keys[i].name = secboot_keys[i].config;
+			printf("\tSuccessfully parsed %s input\n", secboot_keys[i].name);
+			continue;
+		}
+
+		secboot_keys[i].name = strtok(strdup(secboot_keys[i].config), ":");
+
+		if (!secboot_keys[i].name)
+			return -EINVAL;
+
+		handle_string = strtok(NULL, ":");
+		if (!handle_string || strlen(handle_string) != 8)
+			return -EINVAL;
+
+		secboot_keys[i].handle = (uint32_t)simple_strtoul(handle_string, NULL, 16);
+		if (!secboot_keys[i].handle)
+			return -EINVAL;
+		printf("\tSuccessfully parsed %s input\n", secboot_keys[i].name);
+	}
+	return 0;
+}
+
+int hse_check_prev_import(int key_idx)
+{
+	for (int i = 0; i < key_idx; i++) {
+		if (secboot_keys[i].handle == secboot_keys[key_idx].handle) {
+			printf("\tKey slot already occupied, %s already imported for a previous BL\n",
+				 secboot_keys[key_idx].name);
+			return 1;
+		}
+	}
+	return 0;
+}
+
+int hse_extract_file(const char *file_name, uint8_t *buf, u32 *plen)
+{
+	uint8_t *tmp = NULL;
+	long long len_read;
+	char devpart[4];
+	loff_t ret = 0;
+
+	if (!file_name || !buf) {
+		printf("ERROR: param empty, please check it\n");
+		return -1;
+	}
+
+	snprintf(devpart, sizeof(devpart), "%s:%d", env_get("mmcdev"), MMC_ROOT_PARTITION);
+	ret = fs_set_blk_dev("mmc", devpart, FS_TYPE_EXT);
+	if (ret) {
+		printf("ERROR: could not set block device\n");
+	}
+
+	tmp = malloc(MAX_BIN_SIZE);
+	if (!tmp) {
+		printf("ERROR: could not allocate space for temporary buffer \n");
+		return -ENOMEM;;
+	}
+
+	ret = fs_read(file_name, (uintptr_t)tmp, 0, 0, &len_read);
+	if (ret < 0) {
+		printf("ERROR: could not read file %s\n", file_name);
+		goto out_free_tmp_buf;
+	}
+
+	memcpy(buf, tmp, len_read);
+	if (plen) {
+		*plen = len_read;
+	}
+
+out_free_tmp_buf:
+	free(tmp);
+
+	return ret;
+}
+
+int get_env_bins(struct hse_private *priv)
+{
+	/* env bins include:
+	 * required: bl2_bin, bl2_sign
+	*/
+	int ret = 0;
+	char file_name[256];
+	char *file_dir = SECBOOT_RESOURCE_DIR;
+	char *value = NULL;
+
+	value = env_get("bl2_bin");
+	if (!value || !(*value)) {
+		printf("ERROR: failed to get bl2 binary file name\n");
+		return -ENOENT;
+	}
+
+	snprintf(file_name, sizeof(file_name), "/%s/%s", file_dir, value);
+	ret = hse_extract_file(file_name, priv->bl2_bin, &priv->bl2_bin_size);
+	if (ret) {
+	        printf("ERROR: extract BL2 binary file failed\n");
+	        return ret;
+	}
+
+	/* get bl2 sign*/
+	value = env_get("bl2_sign");
+	if (!value || !(*value)) {
+		printf("ERROR: failed to get bl2 signature file name\n");
+		return -ENOENT;
+	}
+	snprintf(file_name, sizeof(file_name), "%s/%s", file_dir, value);
+	ret = hse_extract_file(file_name, priv->bl2_sign, NULL);
+	if (ret) {
+	        printf("ERROR: extract BL2 singnature file failed\n");
+	        return ret;
+	}
+
+	return ret;
+}
+
 /* return 0 for equal uuids */
 static inline int compare_uuids(const struct uuid *uuid1,
 				const struct uuid *uuid2)
@@ -269,7 +420,7 @@ static u32 get_fip_sign_from_dev(struct hse_private *priv)
 	return get_fip_start(priv) + sign_offset;
 }
 
-static int get_m7_sign_data(struct hse_private *priv, bool qspi_boot)
+static int get_m7_sign_data(struct hse_private *priv, enum medium_type boot_medium)
 {
 	u64 sign_addr;
 	int ret = 0;
@@ -283,7 +434,7 @@ static int get_m7_sign_data(struct hse_private *priv, bool qspi_boot)
 		case NXP_PARALLEL_SEC_BOOT:
 			offset = priv->ivt.app_boot;
 			length = sizeof(priv->buf_space);
-			ret = hse_read(offset, length, (void *)&priv->buf_space, qspi_boot);
+			ret = hse_read(offset, length, (void *)&priv->buf_space, boot_medium);
 			if (ret) {
 				printf("ERROR: M7 signature read failed!\n");
 				return ret;
@@ -292,7 +443,7 @@ static int get_m7_sign_data(struct hse_private *priv, bool qspi_boot)
 						M7_BOOT_CODE_SIZE - M7_SIGNATURE_SIZE; 
 			break;
 		case APTIV_AUTOSAR_SEC_BOOT:
-			if (qspi_boot) {
+			if (boot_medium == QSPI) {
 				offset = priv->ivt.app_boot + APP_CODE_OFFSET + priv->app_boot_hdr.code_len;
 				length = M7_SIGNATURE_SIZE;
 			} else {
@@ -300,19 +451,19 @@ static int get_m7_sign_data(struct hse_private *priv, bool qspi_boot)
 				length = SD_SECTOR_SIZE;
 			}
 
-			ret = hse_read(offset, length, (void *)&priv->buf_space, qspi_boot);
+			ret = hse_read(offset, length, (void *)&priv->buf_space, boot_medium);
 			if (ret) {
 				printf("ERROR: M7 signature read failed!\n");
 				return ret;
 			}
 
-			if (qspi_boot)
+			if (boot_medium == QSPI)
 				sign_addr = (uintptr_t)priv->buf_space;
 			else
 				sign_addr = (u64)get_image_start((uintptr_t)priv->buf_space, APP_CODE_OFFSET);
 			break;
 		case APTIV_PARALLEL_SEC_BOOT:
-			if (qspi_boot) {
+			if (boot_medium == QSPI) {
 				offset = priv->ivt.autosar_image_off + APP_CODE_OFFSET + priv->app_boot_hdr.code_len;
 				length = M7_SIGNATURE_SIZE;
 			} else {
@@ -320,13 +471,13 @@ static int get_m7_sign_data(struct hse_private *priv, bool qspi_boot)
 				length = SD_SECTOR_SIZE;
 			}
 
-			ret = hse_read(offset, length, (void *)&priv->buf_space, qspi_boot);
+			ret = hse_read(offset, length, (void *)&priv->buf_space, boot_medium);
 			if (ret) {
 				printf("ERROR: M7 signature read failed!\n");
 				return ret;
 			}
 
-			if (qspi_boot)
+			if (boot_medium == QSPI)
 				sign_addr = (uintptr_t)priv->buf_space;
 			else
 				sign_addr = (u64)get_image_start((uintptr_t)priv->buf_space, APP_CODE_OFFSET);
@@ -336,12 +487,12 @@ static int get_m7_sign_data(struct hse_private *priv, bool qspi_boot)
 			return -1;
 	}
 
-	memcpy((void *)priv->m7_signature, (void *)sign_addr, M7_SIGNATURE_SIZE);
+	memcpy((void *)priv->m7_sign, (void *)sign_addr, M7_SIGNATURE_SIZE);
 
 	return 0;
 }
 
-static int get_fip_sign_data(struct hse_private *priv, bool qspi_boot)
+static int get_fip_sign_data(struct hse_private *priv, enum medium_type boot_medium)
 {
 	uintptr_t sign_offset = 0;;
 	u64 sign_addr;
@@ -355,7 +506,7 @@ static int get_fip_sign_data(struct hse_private *priv, bool qspi_boot)
 	if (priv->ivt.boot_type < APTIV_AUTOSAR_SEC_BOOT) {
 		offset = priv->ivt.app_boot;
 		length = sizeof(priv->buf_space);
-		ret = hse_read(offset, length, (void *)&priv->buf_space, qspi_boot);
+		ret = hse_read(offset, length, (void *)&priv->buf_space, boot_medium);
 		if (ret) {
 				printf("ERROR: FIP signature read failed!\n");
 				return ret;
@@ -369,7 +520,7 @@ static int get_fip_sign_data(struct hse_private *priv, bool qspi_boot)
 	} else {
 		offset = priv->ivt.fip_image_off;
 		length = sizeof(priv->buf_space);
-		ret = hse_read(offset, length, (void *)&priv->buf_space, qspi_boot);
+		ret = hse_read(offset, length, (void *)&priv->buf_space, boot_medium);
 		if (ret) {
 			printf("ERROR: FIP signature read failed!\n");
 			return ret;
@@ -400,23 +551,23 @@ static int get_fip_sign_data(struct hse_private *priv, bool qspi_boot)
 			return -1;
 	}
 
-	memcpy((void *)priv->fip_signature, (void *)sign_addr, HSE_FIP_AUTH_LEN);
+	memcpy((void *)priv->bl2_sign, (void *)sign_addr, HSE_APP_AUTH_LEN);
 
 	return 0;
 }
 
-static int get_signature(struct hse_private *priv, bool qspi_boot)
+static int get_signature(struct hse_private *priv, enum medium_type boot_medium)
 {
 	int ret = 0;
 
-	ret = get_fip_sign_data(priv, qspi_boot);
+	ret = get_fip_sign_data(priv, boot_medium);
 	if (ret) {
 		printf("ERROR: get fip signature data failed!\n");
 		return ret;
 	}
 
 	if (priv->ivt.boot_type > NXP_SEC_BOOT) {
-		ret = get_m7_sign_data(priv, qspi_boot);
+		ret = get_m7_sign_data(priv, boot_medium);
 		if (ret) {
 			printf("ERROR: get m7 signature data failed!\n");
 			return ret;
@@ -458,7 +609,7 @@ int hse_format_key_store(struct hse_private *priv, u32 *recv_buf)
 	return 0;
 }
 
-int hse_import_key(struct hse_private *priv, u32 *recv_buf)
+int hse_import_key(struct hse_private *priv, u32 *recv_buf, u32 handle)
 {
 	struct hse_key_info *key_info;
 	struct hse_import_key_srv *import_key_req;
@@ -479,7 +630,7 @@ int hse_import_key(struct hse_private *priv, u32 *recv_buf)
 	key_info->pub_exponent_size = ARRAY_SIZE(priv->rsa_exponent);
 
 	priv->srv_desc.srv_id = HSE_SRV_ID_IMPORT_KEY;
-	import_key_req->key_handle = HSE_BOOT_KEY_HANDLE;
+	import_key_req->key_handle = handle;
 	import_key_req->key_info_addr = (uintptr_t)key_info;
 	import_key_req->key_addr[0] = (uintptr_t)priv->rsa_modulus;
 	import_key_req->key_addr[1] = (uintptr_t)priv->rsa_exponent;
@@ -522,7 +673,7 @@ int hse_install_cr_entry(struct hse_private *priv, u32 *recv_buf)
 	return 0;
 }
 
-int hse_install_fip_cr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_boot)
+int hse_install_fip_cr_entry(struct hse_private *priv, u32 *recv_buf, enum medium_type boot_medium)
 {
 	int ret;
 	u32 offset;
@@ -546,11 +697,11 @@ int hse_install_fip_cr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_
 		case APTIV_PARALLEL_SEC_BOOT:
 			memset((void *)(uintptr_t)priv->buf_space, 0, sizeof(priv->buf_space));
 			offset = priv->ivt.fip_image_off;
-			if (qspi_boot)
+			if (boot_medium == QSPI)
 				length = sizeof(struct app_boot_hdr);
 			else
 				length = SD_SECTOR_SIZE;
-			ret = hse_read(offset, length, (void *)&priv->buf_space, qspi_boot);
+			ret = hse_read(offset, length, (void *)&priv->buf_space, boot_medium);
 			if (ret) {
 				printf("ERROR: FIP boot app header read failed!\n");
 				return ret;
@@ -587,7 +738,7 @@ int hse_install_fip_cr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_
 	return ret;
 }
 
-int hse_install_m7_cr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_boot)
+int hse_install_m7_cr_entry(struct hse_private *priv, u32 *recv_buf, enum medium_type boot_medium)
 {
 	int ret;
 
@@ -634,14 +785,14 @@ int hse_install_m7_cr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_b
 	return ret;
 }
 
-int hse_install_cr_entries(struct hse_private *priv, u32 *recv_buf, bool qspi_boot)
+int hse_install_cr_entries(struct hse_private *priv, u32 *recv_buf, enum medium_type boot_medium)
 {
 	int ret = 0;
 
 	if (priv->ivt.boot_type == NXP_SEC_BOOT ||
 		priv->ivt.boot_type == NXP_PARALLEL_SEC_BOOT || 
 		priv->ivt.boot_type == APTIV_PARALLEL_SEC_BOOT ) {
-		ret = hse_install_fip_cr_entry(priv, recv_buf, qspi_boot);
+		ret = hse_install_fip_cr_entry(priv, recv_buf, boot_medium);
 		if (ret) {
 			printf("ERROR: fip core reset entry install failed!\n");
 			return ret;
@@ -649,7 +800,7 @@ int hse_install_cr_entries(struct hse_private *priv, u32 *recv_buf, bool qspi_bo
 	}
 
 	if (priv->ivt.boot_type > NXP_SEC_BOOT) {
-		ret = hse_install_m7_cr_entry(priv, recv_buf, qspi_boot);
+		ret = hse_install_m7_cr_entry(priv, recv_buf, boot_medium);
 		if (ret) {
 			printf("ERROR: m7 core reset entry install failed!\n");
 			return ret;
@@ -660,7 +811,7 @@ int hse_install_cr_entries(struct hse_private *priv, u32 *recv_buf, bool qspi_bo
 }
 
 int hse_install_smr_entry(struct hse_private *priv, u32 *recv_buf,
-								struct hse_smr_dynamic_config *dynamic_config, bool qspi_boot)
+								struct hse_smr_dynamic_config *dynamic_config, u32 auth_handle, enum medium_type boot_medium)
 {
 	struct hse_smr_entry *smr_entry;
 	struct hse_smr_install_srv *smr_install_req;
@@ -675,17 +826,18 @@ int hse_install_smr_entry(struct hse_private *priv, u32 *recv_buf,
 	smr_entry->smr_src = dynamic_config->entry_smr_src;
 	smr_entry->smr_dst_addr = dynamic_config->entry_smr_dst_addr;
 	smr_entry->smr_size = dynamic_config->entry_smr_size;
-	if (qspi_boot)
-		smr_entry->config_flags = (HSE_SMR_CFG_FLAG_QSPI_FLASH |
-					   HSE_SMR_CFG_FLAG_INSTALL_AUTH);
-	else
-		smr_entry->config_flags = (HSE_SMR_CFG_FLAG_SD_FLASH |
-					   HSE_SMR_CFG_FLAG_INSTALL_AUTH);
+
+	if (boot_medium == QSPI)
+		smr_entry->config_flags = (HSE_SMR_CFG_FLAG_QSPI_FLASH);
+	else if (boot_medium == SD)
+		smr_entry->config_flags = (HSE_SMR_CFG_FLAG_SD_FLASH);
+	else if (boot_medium == EMMC)
+		smr_entry->config_flags = (HSE_SMR_CFG_FLAG_MMC_FLASH);
+
 	smr_entry->check_period = 0;
-	smr_entry->key_handle = HSE_BOOT_KEY_HANDLE;
+	smr_entry->key_handle = auth_handle;
 	smr_entry->sign_sch.sign_scheme = HSE_SIGN_RSASSA_PKCS1_V15;
 	smr_entry->sign_sch.sch.hash_algo = HSE_HASH_ALGO_SHA_1;
-	smr_entry->auth_tag = dynamic_config->entry_auth_tag;
 	smr_entry->decrypt_key_handle = HSE_SMR_DECRYPT_KEY_HANDLE_NOT_USED;
 	smr_entry->version_offset = 0;
 
@@ -712,9 +864,8 @@ int hse_install_smr_entry(struct hse_private *priv, u32 *recv_buf,
 	return 0;
 }
 
-int hse_install_fip_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_boot)
+int hse_install_app_smr_entry(struct hse_private *priv, u32 *recv_buf, enum medium_type boot_medium)
 {
-	u32 fip_size;
 	int ret = 0;
 	u32 offset;
 	size_t length;
@@ -731,49 +882,40 @@ int hse_install_fip_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi
 		offset =  priv->ivt.fip_image_off;
 	length = sizeof(priv->buf_space);
 
-	ret = hse_read(offset, length, (void *)&priv->buf_space, qspi_boot);
+	ret = hse_read(offset, length, (void *)&priv->buf_space, boot_medium);
 	if (ret) {
-			printf("ERROR: fip code read failed!\n");
-			return ret;
-	}
-
-	fip_size = get_fip_size(priv);
-	if (!fip_size) {
-		printf("ERROR: invalid FIP size!\n");
-		return -ENOMEM;
+		printf("ERROR: app code read failed!\n");
+		return ret;
 	}
 
 	switch (priv->ivt.boot_type) {
 		case NXP_SEC_BOOT:
 			smr_dynamic_config.entry_smr_src = get_image_start((uintptr_t)priv->ivt.app_boot, APP_CODE_OFFSET);
 			smr_dynamic_config.entry_smr_dst_addr = priv->app_boot_hdr.ram_load;
-			smr_dynamic_config.entry_smr_size = fip_size - HSE_FIP_AUTH_LEN;
-			smr_dynamic_config.entry_auth_tag = get_fip_sign_from_dev(priv);
+			smr_dynamic_config.entry_smr_size = priv->bl2_bin_size;
 
-			smr_dynamic_config.req_smr_data_addr = (u64)get_image_start((uintptr_t)priv->buf_space, APP_CODE_OFFSET);
-			smr_dynamic_config.req_smr_data_len = fip_size - HSE_FIP_AUTH_LEN;
+			smr_dynamic_config.req_smr_data_addr = (u64)priv->bl2_bin;
+			smr_dynamic_config.req_smr_data_len = priv->bl2_bin_size;
 			smr_dynamic_config.req_entry_index = 1u;
 			break;
 		case NXP_M7_SEC_BOOT:
 		case NXP_PARALLEL_SEC_BOOT:
 			smr_dynamic_config.entry_smr_src = get_image_start((uintptr_t)priv->ivt.app_boot, APP_CODE_OFFSET) + M7_BOOT_CODE_SIZE;
 			smr_dynamic_config.entry_smr_dst_addr = priv->app_boot_hdr.ram_load + M7_BOOT_CODE_SIZE;
-			smr_dynamic_config.entry_smr_size = fip_size - HSE_FIP_AUTH_LEN;
-			smr_dynamic_config.entry_auth_tag = get_fip_sign_from_dev(priv);
+			smr_dynamic_config.entry_smr_size = priv->bl2_bin_size;
 
-			smr_dynamic_config.req_smr_data_addr = get_image_start((uintptr_t)priv->buf_space, APP_CODE_OFFSET) + M7_BOOT_CODE_SIZE;
-			smr_dynamic_config.req_smr_data_len = fip_size - HSE_FIP_AUTH_LEN;
+			smr_dynamic_config.req_smr_data_addr = (u64)priv->bl2_bin;
+			smr_dynamic_config.req_smr_data_len = priv->bl2_bin_size;
 			smr_dynamic_config.req_entry_index = 2u;
 			break;
 		case APTIV_AUTOSAR_SEC_BOOT:
 		case APTIV_PARALLEL_SEC_BOOT:
 			smr_dynamic_config.entry_smr_src =  get_image_start((uintptr_t)priv->ivt.fip_image_off, APP_CODE_OFFSET);
 			smr_dynamic_config.entry_smr_dst_addr =((struct app_boot_hdr *)priv->buf_space)->ram_load;
-			smr_dynamic_config.entry_smr_size = fip_size;
-			smr_dynamic_config.entry_auth_tag = get_image_start((uintptr_t)priv->ivt.fip_image_off, APP_CODE_OFFSET) + fip_size;
+			smr_dynamic_config.entry_smr_size = priv->bl2_bin_size;
 
 			smr_dynamic_config.req_smr_data_addr = (u64)get_image_start((uintptr_t)priv->buf_space, APP_CODE_OFFSET);
-			smr_dynamic_config.req_smr_data_len = fip_size;
+			smr_dynamic_config.req_smr_data_len = priv->bl2_bin_size;
 			smr_dynamic_config.req_entry_index = 2u;
 			break;
 		default:
@@ -781,10 +923,10 @@ int hse_install_fip_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi
 			return -1;
 	}
 
-	smr_dynamic_config.req_smr_auth_tag_addr = (uintptr_t)priv->fip_signature;
-	smr_dynamic_config.req_smr_auth_tag_len = HSE_FIP_AUTH_LEN;
+	smr_dynamic_config.req_smr_auth_tag_addr = (uintptr_t)priv->bl2_sign;
+	smr_dynamic_config.req_smr_auth_tag_len = HSE_APP_AUTH_LEN;
 
-	ret = hse_install_smr_entry(priv, recv_buf, &smr_dynamic_config, qspi_boot);
+	ret = hse_install_smr_entry(priv, recv_buf, &smr_dynamic_config, secboot_keys[BL2_KEY].handle, boot_medium);
 	if (ret) {
 		printf("ERROR: install FIP smr failed!\n");
 		return ret;
@@ -793,7 +935,7 @@ int hse_install_fip_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi
 	return ret;
 }
 
-int hse_install_m7_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_boot)
+int hse_install_m7_smr_entry(struct hse_private *priv, u32 *recv_buf, enum medium_type boot_medium)
 {
 	u32 align_diff;
 	int ret = 0;
@@ -814,14 +956,14 @@ int hse_install_m7_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_
 			offset = priv->ivt.app_boot;
 		else
 			offset = priv->ivt.autosar_image_off;
-		if (qspi_boot) {
+		if (boot_medium == QSPI) {
 			length = priv->app_boot_hdr.code_len + APP_CODE_OFFSET;
 		} else {
 			length = priv->app_boot_hdr.code_len + SD_SECTOR_SIZE;
 		}
 	}
 
-	ret = hse_read(offset, length, (void *)&priv->buf_space, qspi_boot);
+	ret = hse_read(offset, length, (void *)&priv->buf_space, boot_medium);
 	if (ret) {
 		printf("ERROR: m7 code read failed!\n");
 		return ret;
@@ -837,7 +979,6 @@ int hse_install_m7_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_
 			smr_dynamic_config.entry_smr_src = get_image_start((uintptr_t)priv->ivt.app_boot, APP_CODE_OFFSET) + align_diff;
 			smr_dynamic_config.entry_smr_dst_addr = priv->app_boot_hdr.ram_entry;
 			smr_dynamic_config.entry_smr_size = priv->ivt.m7_image_size;
-			smr_dynamic_config.entry_auth_tag = get_m7_sign_from_dev(priv);
 
 			smr_dynamic_config.req_smr_data_addr = get_image_start((uintptr_t)priv->buf_space, APP_CODE_OFFSET) + align_diff;
 			smr_dynamic_config.req_smr_data_len = priv->ivt.m7_image_size;
@@ -851,7 +992,6 @@ int hse_install_m7_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_
 			smr_dynamic_config.entry_smr_src = get_image_start((uintptr_t)priv->ivt.app_boot, APP_CODE_OFFSET) + align_diff;
 			smr_dynamic_config.entry_smr_dst_addr = priv->ivt.m7_image_entry_bak;
 			smr_dynamic_config.entry_smr_size = priv->ivt.m7_image_size;
-			smr_dynamic_config.entry_auth_tag = get_m7_sign_from_dev(priv);
 
 			smr_dynamic_config.req_smr_data_addr = get_image_start((uintptr_t)priv->buf_space, APP_CODE_OFFSET) + align_diff;
 			smr_dynamic_config.req_smr_data_len = priv->ivt.m7_image_size;
@@ -860,7 +1000,6 @@ int hse_install_m7_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_
 			smr_dynamic_config.entry_smr_src = get_image_start((uintptr_t)priv->ivt.app_boot, APP_CODE_OFFSET);
 			smr_dynamic_config.entry_smr_dst_addr = priv->app_boot_hdr.ram_load;
 			smr_dynamic_config.entry_smr_size = priv->app_boot_hdr.code_len;
-			smr_dynamic_config.entry_auth_tag = get_m7_sign_from_dev(priv);
 
 			smr_dynamic_config.req_smr_data_addr = (u64)get_image_start((uintptr_t)priv->buf_space, APP_CODE_OFFSET);
 			smr_dynamic_config.req_smr_data_len = priv->app_boot_hdr.code_len;
@@ -869,7 +1008,6 @@ int hse_install_m7_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_
 			smr_dynamic_config.entry_smr_src = get_image_start((uintptr_t)priv->ivt.autosar_image_off, APP_CODE_OFFSET);
 			smr_dynamic_config.entry_smr_dst_addr = priv->app_boot_hdr.ram_load;
 			smr_dynamic_config.entry_smr_size = priv->app_boot_hdr.code_len;
-			smr_dynamic_config.entry_auth_tag = get_m7_sign_from_dev(priv);
 
 			smr_dynamic_config.req_smr_data_addr = (u64)get_image_start((uintptr_t)priv->buf_space, APP_CODE_OFFSET);
 			smr_dynamic_config.req_smr_data_len = priv->app_boot_hdr.code_len;
@@ -880,10 +1018,10 @@ int hse_install_m7_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_
 	}
 
 	smr_dynamic_config.req_entry_index = 1u;
-	smr_dynamic_config.req_smr_auth_tag_addr = (u64)priv->m7_signature;
+	smr_dynamic_config.req_smr_auth_tag_addr = (u64)priv->m7_sign;
 	smr_dynamic_config.req_smr_auth_tag_len = M7_SIGNATURE_SIZE;
 
-	ret = hse_install_smr_entry(priv, recv_buf, &smr_dynamic_config, qspi_boot);
+	ret = hse_install_smr_entry(priv, recv_buf, &smr_dynamic_config, DEFAULT_KEY_HANDLE, boot_medium);
 	if (ret) {
 		printf("ERROR: install M7 smr failed!\n");
 		return ret;
@@ -892,18 +1030,18 @@ int hse_install_m7_smr_entry(struct hse_private *priv, u32 *recv_buf, bool qspi_
 	return ret;
 }
 
-int hse_install_smr_entries(struct hse_private *priv, u32 *recv_buf, bool qspi_boot)
+int hse_install_smr_entries(struct hse_private *priv, u32 *recv_buf, enum medium_type boot_medium)
 {
 	int ret = 0;
 
-	ret = hse_install_fip_smr_entry(priv, recv_buf, qspi_boot);
+	ret = hse_install_app_smr_entry(priv, recv_buf, boot_medium);
 	if (ret) {
 		printf("ERROR: fip smr entry install failed!\n");
 		return ret;
 	}
 
 	if (priv->ivt.boot_type > NXP_SEC_BOOT) {
-		ret = hse_install_m7_smr_entry(priv, recv_buf, qspi_boot);
+		ret = hse_install_m7_smr_entry(priv, recv_buf, boot_medium);
 		if (ret) {
 			printf("ERROR: m7 smr entry install failed!\n");
 			return ret;
@@ -963,7 +1101,7 @@ static ulong bytes_per_second(unsigned int len, ulong start_ms)
 		return 1024 * len / max(get_timer(start_ms), 1UL);
 }
 
-int hse_write_sys_img(struct hse_private *priv, bool secure, bool qspi_boot)
+int hse_write_sys_img(struct hse_private *priv, bool secure, enum medium_type boot_medium)
 {
 	int ret;
 	u32 offset;
@@ -976,19 +1114,18 @@ int hse_write_sys_img(struct hse_private *priv, bool secure, bool qspi_boot)
 
 	offset = priv->ivt.sys_img;
 	length = HSE_SYS_IMG_MAX_SIZE;
-	ret = hse_write(offset, length, (char *)&priv->sys_img, qspi_boot);
+	ret = hse_write(offset, length, (char *)&priv->sys_img, boot_medium);
 	if (ret) {
 		printf("ERROR: sys-img publish failed during writing data!\n");
 		return ret;
 	}
 
-	if (qspi_boot) {
+	priv->ivt.sys_img_ext_flash_type = boot_medium;
+	if (boot_medium == QSPI) {
 		/* external flash type, flash page size */
-		priv->ivt.sys_img_ext_flash_type = HSE_EXT_FLASH_QSPI;
 		priv->ivt.sys_img_flash_page_size = HSE_EXT_FLASH_PAGE_QSPI;
 	} else {
 		/* external flash type, flash page size */
-		priv->ivt.sys_img_ext_flash_type = HSE_EXT_FLASH_SD;
 		priv->ivt.sys_img_flash_page_size = HSE_EXT_FLASH_PAGE;
 	}
 
@@ -997,14 +1134,14 @@ int hse_write_sys_img(struct hse_private *priv, bool secure, bool qspi_boot)
 		priv->ivt.boot_cfg |= HSE_IVT_BOOTSEQ_BIT;
 
 	/* write ivt */
-	if (qspi_boot) {
+	if (boot_medium == QSPI) {
 		offset = HSE_IVT_QSPI_OFFSET;
 		length = sizeof(priv->ivt);
 	} else {
 		offset = HSE_IVT_BLK * SD_SECTOR_SIZE;
 		length = SD_SECTOR_SIZE;
 	}
-	ret = hse_write(offset, length, (char *)&priv->ivt, qspi_boot);
+	ret = hse_write(offset, length, (char *)&priv->ivt, boot_medium);
 	if (ret) {
 		printf("ERROR: ivt write failed!\n");
 		return ret;
@@ -1058,13 +1195,13 @@ int hse_enable_mus(struct hse_private *priv, u32 *recv_buf)
 	return 0;
 }
 
-static int get_ivt(struct hse_private *priv, bool qspi_boot)
+static int get_ivt(struct hse_private *priv, enum medium_type boot_medium)
 {
 	int ret = 0;
 	u32 offset;
 	size_t length;
 
-	if (qspi_boot) {
+	if (boot_medium == QSPI) {
 		offset = HSE_IVT_QSPI_OFFSET;
 		length = sizeof(priv->ivt);
 	} else {
@@ -1072,12 +1209,12 @@ static int get_ivt(struct hse_private *priv, bool qspi_boot)
 		length = SD_SECTOR_SIZE;
 	}
 
-	ret = hse_read(offset, length, (void *)&priv->ivt, qspi_boot);
+	ret = hse_read(offset, length, (void *)&priv->ivt, boot_medium);
 
 	return ret;
 }
 
-static int get_app_header(struct hse_private *priv, bool qspi_boot)
+static int get_app_header(struct hse_private *priv, enum medium_type boot_medium)
 {
 	int ret = 0;
 	u32 offset;
@@ -1089,31 +1226,34 @@ static int get_app_header(struct hse_private *priv, bool qspi_boot)
 		offset = priv->ivt.autosar_image_off;
 	}
 
-	if (qspi_boot) {
+	if (boot_medium == QSPI) {
 		length = sizeof(priv->app_boot_hdr);
 	} else {
 		length = SD_SECTOR_SIZE;
 	}
 
-	ret = hse_read(offset, length, (void *)&priv->app_boot_hdr, qspi_boot);
+	ret = hse_read(offset, length, (void *)&priv->app_boot_hdr, boot_medium);
 
 	return ret;
 }
 
-static int get_public_key(struct hse_private *priv, char *pubkey_file)
+static int get_public_key(struct hse_private *priv, struct bl_key *key)
 {
-	char mmcdevpart[4];
+	char devpart[4];
 	long long len_read;
+	char *file_dir = SECBOOT_RESOURCE_DIR;
+	char file_name[256];
 	int ret = 0;
 
-	snprintf(mmcdevpart, sizeof(mmcdevpart), "%s:%s",
-		 env_get("mmcdev"), env_get("mmcpart"));
-	ret = fs_set_blk_dev("mmc", mmcdevpart, FS_TYPE_FAT);
+	snprintf(devpart, sizeof(devpart), "%s:%d", env_get("mmcdev"), MMC_ROOT_PARTITION);
+	ret = fs_set_blk_dev("mmc", devpart, FS_TYPE_EXT);
 	if (ret) {
 		printf("ERROR: could not set block device!\n");
 		return ret;
 	}
-	ret = fs_read(pubkey_file, (uintptr_t)priv->rsa_pubkey, 0, 0, &len_read);
+
+	snprintf(file_name, sizeof(file_name), "/%s/%s", file_dir, key->name);
+	ret = fs_read(file_name, (uintptr_t)priv->rsa_pubkey, 0, 0, &len_read);
 	if (ret < 0) {
 		printf("ERROR: could not read public key file!\n");
 		return ret;
@@ -1129,6 +1269,34 @@ static int get_public_key(struct hse_private *priv, char *pubkey_file)
 	return ret;
 }
 
+int hse_import_keys(struct hse_private *priv)
+{
+	int ret = 0;
+	u32 hse_recv;
+
+	for (int i = 0; secboot_keys[i].type != INVALID_KEY; i++) {
+		if (!secboot_keys[i].name || hse_check_prev_import(i))
+			continue;
+
+		/* read public key file */
+		ret = get_public_key(priv, &secboot_keys[i]);
+		if (ret) {
+			printf("ERROR: get public key failed!\n");
+			return ret;
+		}
+
+		printf("\tImporting %s\n", secboot_keys[i].name);
+		ret = hse_import_key(priv, &hse_recv, secboot_keys[i].handle);
+		if (ret) {
+			printf("ERROR: failed to import %s\n", secboot_keys[i].name);
+			return ret;
+		}
+	}
+
+	return ret;
+
+}
+
 static int setup_qspi_flash(struct spi_flash **flash)
 {
 #if CONFIG_IS_ENABLED(DM_SPI_FLASH)
@@ -1162,12 +1330,11 @@ static int do_hse_secboot_enable(struct cmd_tbl *cmdtp, int flag,
 				 int argc, char * const argv[])
 {
 	struct hse_private *priv;
-	char *pubkey_file;
 	u16 hse_status_ret;
 	u32 hse_recv;
 	u64 hse_resmem;
 	int hse_nodeoffset, ret;
-	bool qspi_boot = false;
+	enum medium_type boot_medium = SD;
 
 	/* check if hse has been initialised */
 	hse_status_ret = hse_mu_check_status();
@@ -1176,24 +1343,27 @@ static int do_hse_secboot_enable(struct cmd_tbl *cmdtp, int flag,
 		return CMD_RET_FAILURE;
 	}
 
-	if (argc < 3 || !argv[2]) {
-		printf("USAGE: hse_secboot_enable <device> <public_key_file>.der\n");
+	if (argc < 2 || !argv[1]) {
+		printf("USAGE: hse_secboot_enable <device>\n");
 		printf("\n");
-		printf("    <device> - boot device: sd/qspi\n");
-		printf("    <public_key_file>.der - rsa public key in DER format\n");
-		printf("                            in the FAT partition\n");
+		printf("    <device> - boot device: sd/emmc/qspi\n");
 		return CMD_RET_FAILURE;
 	}
 
 	if (!strcmp(argv[1], "qspi")) {
-		qspi_boot = true;
+		boot_medium = QSPI;
 		ret = setup_qspi_flash(&qspi_flash);
 		if (ret)
 			return ret;
+	} else if (!strcmp(argv[1], "emmc")) {
+		boot_medium = EMMC;
+	} else if (!strcmp(argv[1], "sd")) {
+		boot_medium = SD;
+	} else {
+		printf("Unknown boot device\n");
+		return CMD_RET_FAILURE;
 	}
 
-	pubkey_file = argv[2];
-
 	/* find mem reserved for hse */
 	hse_nodeoffset = fdt_node_offset_by_compatible(gd->fdt_blob, -1,
 						       "nxp,s32cc-hse-rmem");
@@ -1212,40 +1382,26 @@ static int do_hse_secboot_enable(struct cmd_tbl *cmdtp, int flag,
 	memset((void *)priv, 0, sizeof(struct hse_private));
 
 	/* read ivt block */
-	ret = get_ivt(priv, qspi_boot);
+	ret = get_ivt(priv, boot_medium);
 	if (ret) {
 		printf("ERROR: ivt read failed!\n");
 		return ret;
 	}
 
 	/* read app boot code header */
-	ret = get_app_header(priv, qspi_boot);
+	ret = get_app_header(priv, boot_medium);
 	if (ret) {
 		printf("ERROR: app boot code header read failed!\n");
 		return ret;
 	}
 
-	/* get signature data */
-	ret = get_signature(priv, qspi_boot);
-	if (ret) {
-		printf("ERROR: get signature data failed!\n");
-		return ret;
-	}
-
-	/* read public key file */
-	ret = get_public_key(priv, pubkey_file);
-	if (ret) {
-		printf("ERROR: get public key failed!\n");
-		return ret;
-	}
-
 	/* enable hse mailbox */
 	ret = hse_enable_mus(priv, &hse_recv);
 	if (ret)
 		return ret;
 
 	/* check if sys_img already exists */
-	if (!(hse_status_ret & HSE_STATUS_PRIMARY_SYSIMG)) {
+	if (!(hse_status_ret & HSE_STATUS_INSTALL_OK)) {
 		printf("\tNo SYS_IMG, formatting key store...\n");
 
 		ret = hse_format_key_store(priv, &hse_recv);
@@ -1253,15 +1409,35 @@ static int do_hse_secboot_enable(struct cmd_tbl *cmdtp, int flag,
 			return ret;
 	}
 
-	ret = hse_import_key(priv, &hse_recv);
-	if (ret)
+	ret = get_env_bins(priv);
+	if (ret) {
+		printf("ERROR: get secboot environment binary variables failed!\n");
 		return ret;
+	}
+
+	ret = get_env_keys();
+	if (ret) {
+		printf("ERROR: get secboot environment key variables failed!\n");
+		return ret;
+	}
+
+	ret = hse_parse_keys();
+	if (ret) {
+		printf("ERROR: parse secboot key variables failed!\n");
+		return ret;
+	}
 
-	ret = hse_install_smr_entries(priv, &hse_recv, qspi_boot);
+	ret = hse_import_keys(priv);
+	if (ret) {
+		printf("ERROR: import secboot pubkeys failed!\n");
+		return ret;
+	}
+
+	ret = hse_install_smr_entries(priv, &hse_recv, boot_medium);
 	if (ret)
 		return ret;
 
-	ret = hse_install_cr_entries(priv, &hse_recv, qspi_boot);
+	ret = hse_install_cr_entries(priv, &hse_recv, boot_medium);
 	if (ret)
 		return ret;
 
@@ -1269,14 +1445,14 @@ static int do_hse_secboot_enable(struct cmd_tbl *cmdtp, int flag,
 	if (ret) 
 		return ret;
 
-	ret = hse_write_sys_img(priv, true, qspi_boot);
+	ret = hse_write_sys_img(priv, true, boot_medium);
 	if (ret) 
 		return ret;
 
 	return CMD_RET_SUCCESS;
 }
 
-U_BOOT_CMD(hse_secboot_enable, 3, 0, do_hse_secboot_enable,
+U_BOOT_CMD(hse_secboot_enable, 2, 0, do_hse_secboot_enable,
 	   "generate device-specific SYS_IMG",
 	   "Generate SYS-IMG and place it on SD card");
 
@@ -1288,7 +1464,7 @@ static int do_hse_keystore_format(struct cmd_tbl *cmdtp, int flag,
 	u32 hse_recv;
 	u64 hse_resmem;
 	int hse_nodeoffset, ret;
-	bool qspi_boot = false;
+	enum medium_type boot_medium = SD;
 	u32 offset;
 	size_t length;
 
@@ -1300,10 +1476,16 @@ static int do_hse_keystore_format(struct cmd_tbl *cmdtp, int flag,
 	}
 
 	if (!strcmp(argv[1], "qspi")) {
-		qspi_boot = true;
+		boot_medium = QSPI;
 		ret = setup_qspi_flash(&qspi_flash);
 		if (ret)
 			return ret;
+	} else if (!strcmp(argv[1], "sd")) {
+		boot_medium = SD;
+	} else if (!strcmp(argv[1], "emmc"))	{
+		boot_medium = EMMC;
+	} else {
+		printf("Unknown boot devices\n");
 	}
 
 	/* check if hse has been initialised */
@@ -1314,7 +1496,7 @@ static int do_hse_keystore_format(struct cmd_tbl *cmdtp, int flag,
 	}
 
 	/* check if sys_img already exists */
-	if (hse_status_ret & HSE_STATUS_PRIMARY_SYSIMG) {
+	if (hse_status_ret & HSE_STATUS_INSTALL_OK) {
 		printf("CHECK: SYS_IMG already loaded\n");
 		return CMD_RET_SUCCESS;
 	}
@@ -1337,14 +1519,14 @@ static int do_hse_keystore_format(struct cmd_tbl *cmdtp, int flag,
 	memset((void *)priv, 0, sizeof(struct hse_private));
 
 	/* read ivt */
-	if (qspi_boot) {
+	if (boot_medium == QSPI) {
 		offset = HSE_IVT_QSPI_OFFSET;
 		length = sizeof(priv->ivt);
 	} else {
 		offset = HSE_IVT_BLK  *SD_SECTOR_SIZE;
 		length = SD_SECTOR_SIZE;
 	}
-	ret = hse_read(offset, length, (void *)&priv->ivt, qspi_boot);
+	ret = hse_read(offset, length, (void *)&priv->ivt, boot_medium);
 	if (ret) {
 		printf("ERROR: ivt read failed!\n");
 		return ret;
@@ -1362,7 +1544,7 @@ static int do_hse_keystore_format(struct cmd_tbl *cmdtp, int flag,
 	if (ret)
 		return ret;
 
-	ret = hse_write_sys_img(priv, false, qspi_boot);
+	ret = hse_write_sys_img(priv, false, boot_medium);
 	if (ret)
 		return ret;
 
diff --git a/include/configs/s32-cc.h b/include/configs/s32-cc.h
index c372f3f9..83505fce 100644
--- a/include/configs/s32-cc.h
+++ b/include/configs/s32-cc.h
@@ -130,6 +130,11 @@
 	"fitimage=fitImage\0" \
 	"fitimage_addr=" __stringify(S32CC_FIT_ADDR) "\0" \
 	"fit_config_header=conf-\0" \
+	"bl2_bin=" BL2_BIN_FILE "\0"\
+	"bl2_sign=" BL2_SIGN_FILE "\0"\
+	"bl2_pubkey=bl2_rsa2048_public.der:0x010700\0"\
+	"bl31_pubkey=bl31_rsa2048_public.der:0x010700\0"\
+	"bl33_pubkey=bl33_rsa2048_public.der:0x010700\0"\
 	"ipaddr=" S32CC_IPADDR "\0"\
 	"loadtftpfdt=tftp ${fdt_addr} ${fdt_file};\0" \
 	"loadfitimage=fatload mmc ${mmcdev}:${mmcpart} ${fitimage_addr} ${fitimage}\0" \
diff --git a/include/configs/s32g274ardb2.h b/include/configs/s32g274ardb2.h
index 1a8bf5cc..cb1e728b 100644
--- a/include/configs/s32g274ardb2.h
+++ b/include/configs/s32g274ardb2.h
@@ -9,5 +9,7 @@
 
 #define EXTRA_BOOT_ARGS			""
 #define FDT_FILE			"s32g274a-rdb2.dtb"
+#define BL2_BIN_FILE        "bl2_w_dtb-s32g274ardb2.bin"
+#define BL2_SIGN_FILE       "bl2_w_dtb-s32g274ardb2.bin.sign"
 
 #endif
diff --git a/include/configs/s32g2xxaevb.h b/include/configs/s32g2xxaevb.h
index 0b2bbc33..cad591d7 100644
--- a/include/configs/s32g2xxaevb.h
+++ b/include/configs/s32g2xxaevb.h
@@ -9,6 +9,8 @@
 
 #define EXTRA_BOOT_ARGS			""
 #define FDT_FILE			"s32g2xxa-evb.dtb"
+#define BL2_BIN_FILE        "bl2_w_dtb-s32g2xxaevb.bin"
+#define BL2_SIGN_FILE       "bl2_w_dtb-s32g2xxaevb.bin.sign"
 
 #if defined(CONFIG_USB)
 #	define CONFIG_MXC_USB_PORTSC	PORT_PTS_ULPI
diff --git a/include/configs/s32g399ardb3.h b/include/configs/s32g399ardb3.h
index a75702ed..06a1123a 100644
--- a/include/configs/s32g399ardb3.h
+++ b/include/configs/s32g399ardb3.h
@@ -15,5 +15,7 @@
 #else
 #define FDT_FILE			"s32g399a-rdb3-pfems.dtb"
 #endif
+#define BL2_BIN_FILE        "bl2_w_dtb-s32g399ardb3.bin"
+#define BL2_SIGN_FILE       "bl2_w_dtb-s32g399ardb3.bin.sign"
 
 #endif
diff --git a/include/configs/s32g3xxaevb.h b/include/configs/s32g3xxaevb.h
index 3e7717d5..abc1e5cc 100644
--- a/include/configs/s32g3xxaevb.h
+++ b/include/configs/s32g3xxaevb.h
@@ -9,6 +9,8 @@
 
 #define EXTRA_BOOT_ARGS			""
 #define FDT_FILE			"s32g3xxa-evb.dtb"
+#define BL2_BIN_FILE        "bl2_w_dtb-s32g3xxaevb.bin"
+#define BL2_SIGN_FILE       "bl2_w_dtb-s32g3xxaevb.bin.sign"
 
 #if defined(CONFIG_USB)
 #	define CONFIG_MXC_USB_PORTSC	PORT_PTS_ULPI
diff --git a/include/hse/hse_abi.h b/include/hse/hse_abi.h
index bed0987a..5bea5538 100644
--- a/include/hse/hse_abi.h
+++ b/include/hse/hse_abi.h
@@ -28,7 +28,7 @@
 #define HSE_SRV_RSP_KEY_INVALID   0xA5AA52B4ul
 
 #define HSE_SYS_IMG_MAX_SIZE 0xC000u
-#define HSE_FIP_AUTH_LEN     0x100u
+#define HSE_APP_AUTH_LEN     0x100u
 #define HSE_AUTH_TAG_OFFSET  0x200u
 
 #define HSE_IVT_BLK      8
@@ -47,6 +47,7 @@
 #define HSE_CHANNEL_GENERAL 1u
 
 #define HSE_STATUS_INIT_OK        BIT(8)
+#define HSE_STATUS_INSTALL_OK     BIT(9)
 #define HSE_STATUS_PRIMARY_SYSIMG BIT(14)
 #define HSE_IVT_BOOTSEQ_BIT       BIT(3)
 
@@ -70,6 +71,7 @@
 
 #define HSE_SMR_CFG_FLAG_QSPI_FLASH 0x0u
 #define HSE_SMR_CFG_FLAG_SD_FLASH     0x2u
+#define HSE_SMR_CFG_FLAG_MMC_FLASH     0x3u
 #define HSE_SMR_CFG_FLAG_INSTALL_AUTH BIT(2)
 #define HSE_SMR_ENTRY_1               BIT(1)
 #define HSE_SMR_ENTRY_2               BIT(2)
@@ -120,6 +122,8 @@
 #define HSE_KEY2048_BITS 2048u
 #define HSE_KEY4096_BITS 4096u
 
+#define MAX_BIN_SIZE 	512*1024 /* default binary file's max size set to 512KB */
+
 /* see include/mmc.h */
 struct mmc *hse_init_mmc_device(int dev, bool force_init);
 int hse_mmc_read(void *addr, u32 blk, u32 cnt);
@@ -478,6 +482,8 @@ struct hse_srv_desc {
  * @ram_catalog: ram key catalog
  * @buf_space: buffer used to save tmp data, size 0x150200,
  *             this buffer is used to save the atf or autosar images that are all less than 1.3 MB
+ * @bl2_bin: buffer used to save bl2 bin data, size 512KB.
+ * @bl2_bin_size: the size of the bl2 binary file
  */
 struct hse_private {
 	struct ivt ivt;
@@ -491,14 +497,16 @@ struct hse_private {
 	u8 rsa_modulus[256];
 	u8 rsa_exponent[3];
 	u8 reserved;
-	u8 fip_signature[256];
-	u8 m7_signature[256];
+	u8 bl2_sign[256];
+	u8 m7_sign[256];
 	u8 sys_img[HSE_SYS_IMG_MAX_SIZE];
 	u32 sys_img_len;
 	u32 publish_offset;
 	struct hse_key_group_cfg_entry nvm_catalog[20];
 	struct hse_key_group_cfg_entry ram_catalog[11];
 	u8 buf_space[1376768];
+	u8 bl2_bin[MAX_BIN_SIZE];
+	u32 bl2_bin_size;
 };
 
 enum boot_type {
@@ -511,6 +519,41 @@ enum boot_type {
 	APTIV_PARALLEL_SEC_BOOT,
 };
 
+enum medium_type {
+	QSPI = 0,
+	SD = 2,
+	EMMC = 3,
+};
+
+#define APP_CODE_OFFSET 0x40
+#define M7_BOOT_CODE_SIZE 0x2000
+#define M7_SIGNATURE_SIZE 0x100
+
+#define DEFAULT_KEY_HANDLE 0x010700
+#define BL2_DEFAULT_KEY_HANDLE 0x010700
+#define BL31_DEFAULT_KEY_HANDLE 0x010700
+#define BL32_DEFAULT_KEY_HANDLE 0x010700
+#define BL33_DEFAULT_KEY_HANDLE 0x010700
+#define MMC_ROOT_PARTITION		3	/* now the root part of mmc device is 3 */
+#define SECBOOT_RESOURCE_DIR	"boot/secboot"
+
+enum image_key_type {
+	INVALID_KEY = -1,
+	BL2_KEY = 0,
+	BL31_KEY,
+	BL32_KEY,
+	BL33_KEY,
+};
+
+struct bl_key {
+	enum image_key_type type;
+	bool req;
+	uint32_t handle;
+	char *name;
+	char *env_name;
+	char *config;
+};
+
 struct hse_smr_dynamic_config {
 	u32 entry_smr_src;
 	u64 entry_smr_dst_addr;
-- 
2.34.1

